src/deepLink/deepLinkService.ts

```diff
@@ -0,0 +1,293 @@
+
+	constructor(private readonly container: Container) {
+		this._context = {
+			state: DeepLinkServiceState.Idle,
+		};
+
+		this._disposable = container.uri.onDidReceiveUri(async (event: UriEvent) => {
+			if (event.type === UriTypes.DeepLink && this._context.state === DeepLinkServiceState.Idle) {
+				if (!event.repoId || !event.linkType || !event.uri || !event.remoteUrl) {
+					void window.showErrorMessage(`Error resolving deep link: missing required properties.`);
+					return;
+				}
+
+				if (!Object.values(DeepLinkType).includes(event.linkType)) {
+					void window.showErrorMessage(`Error resolving deep link: unknown link type.`);
+					return;
+				}
+
+				if (event.linkType !== DeepLinkType.Remote && !event.targetId) {
+					void window.showErrorMessage(
+						`Error resolving deep link of type ${event.linkType}: no target id provided.`,
+					);
+					return;
+				}
+
+				this._context = {
+					...this._context,
+					repoId: event.repoId,
+					targetType: event.linkType,
+					uri: event.uri,
+					remoteUrl: event.remoteUrl,
+					targetId: event.targetId,
+				};
+
+				await this.processDeepLink();
+			}
+		});
+	}
+
+	resetContext() {
+		this._context = {
+			state: DeepLinkServiceState.Idle,
+			uri: undefined,
+			repoId: undefined,
+			repo: undefined,
+			remoteUrl: undefined,
+			remote: undefined,
+			targetId: undefined,
+			targetType: undefined,
+			targetSha: undefined,
+		};
+	}
+
+	dispose() {
+		this._disposable.dispose();
+	}
+
+	async getShaForTarget(): Promise<string | undefined> {
+		const { repo, remote, targetType, targetId } = this._context;
+		if (!repo || !remote || targetType === DeepLinkType.Remote || !targetId) {
+			return undefined;
+		}
+
+		if (targetType === DeepLinkType.Branch) {
+			// Form the target branch name using the remote name and branch name
+			const branchName = `${remote.name}/${targetId}`;
+			const branch = await repo.getBranch(branchName);
+			if (branch) {
+				return branch.sha;
+			}
+
+			return undefined;
+		}
+
+		if (targetType === DeepLinkType.Tag) {
+			const tag = await repo.getTag(targetId);
+			if (tag) {
+				return tag.sha;
+			}
+
+			return undefined;
+		}
+
+		if (targetType === DeepLinkType.Commit) {
+			if (await this.container.git.validateReference(repo.path, targetId)) {
+				return targetId;
+			}
+
+			return undefined;
+		}
+
+		return undefined;
+	}
+
+	async processDeepLink(): Promise<void> {
+		let message = '';
+		let matchingRemotes: GitRemote[] = [];
+		let action: DeepLinkServiceAction = DeepLinkServiceAction.DeepLinkEventFired;
+		while (true) {
+			this._context.state = deepLinkStateTransitionTable[this._context.state][action];
+			const { state, repoId, repo, uri, remoteUrl, remote, targetSha, targetType } = this._context;
+			switch (state) {
+				case DeepLinkServiceState.Idle:
+					if (action === DeepLinkServiceAction.DeepLinkResolved) {
+						void window.showInformationMessage(`Deep link resolved: ${uri?.toString()}`);
```

Comment:
If everything worked do we need a notification? Feels like overkill.

Reply:
Done in commit: [Log deep link issues](https://github.com/gitkraken/vscode-gitlens/pull/2471/commits/8e25b9de822b6c0be4ff9fb958d12d02d0ea494b)